# Linear programming with Scipy
## Dr. Tirthajyoti Sarkar

Simple, straight-forward linear programming (LP) problems can also be addressed by Scipy. Prior to 2014, it did not have a LP solver built-in, but it has changed since then. Let’s take a practical factory production problem (borrowed from [this example](https://realpython.com/linear-programming-python/) and slightly changed)

---
## The problem

A factory produces four different products, and that the daily produced amount of the first product is $x_1$, the amount produced of the second product is $x_2$, and so on. The goal is to determine the profit-maximizing daily production amount for each product, with the following constraints,

- The profit per unit of product is 20, 12, 30, and 15 for the first, second, third, and fourth product, respectively.

- Due to manpower constraints, the total number of units produced per day can’t exceed fifty (50).

- For each unit of the first product, three units of the raw material A are consumed. Each unit of the second product requires two units of the raw material A and one unit of the raw material B. Each unit of the third product needs two unit of A and five units of B. Finally, each unit of the fourth product requires three units of B.

- Due to the transportation and storage constraints, the factory can consume up to one hundred units of the raw material A and ninety units of B per day.

The linear programming (LP) problem is,

$$\text{maximize: }\ 20x_1+12x_2+30x_3+15x_4$$

$$\text{s.t.: } x_1+x_2+x_3+x_4 \leq 50 \text { (manpower constraint)}$$

$$3x_1+2x_2+2x_3 \leq 100 \text { (material A constraint)}$$

$$x_2+5x_3+3x_4 \leq 90 \text { (material B constraint)}$$

$$x_1, x_2, x_3, x_4 \geq 0$$

---

### Negative coefficients for the objective function because it is a maximization

In [7]:
obj = [-20, -12, -30, -15] 

### Inequality matrices

In [13]:
# LHS matrix of inequality equations
lhs = [[1, 1, 1, 1],
 [3, 2, 2, 0],
 [0, 1, 5, 3]]

In [14]:
# RHS matrix of inequality equations
rhs = [50,
      100,
      90]

### Setup and solve in Scipy

In [15]:
from scipy.optimize import linprog

In [16]:
lp_opt = linprog(c=obj,
                 A_ub=lhs,
                 b_ub=rhs,
                 method = 'interior-point')

In [17]:
lp_opt

     con: array([], dtype=float64)
     fun: -1033.3333113034805
 message: 'Optimization terminated successfully.'
     nit: 4
   slack: array([1.06529068e-06, 2.14344466e-06, 1.86209118e-06])
  status: 0
 success: True
       x: array([2.66666661e+01, 1.84050438e-08, 9.99999980e+00, 1.33333330e+01])

## Note

So, the solution says that,

- The factory should produce 26.66 units of $x_1$, 10 units of $x_3$, and 13.33 units of $x_4$ every day. The extremely small number corresponding to $x_2$ essentially indicates that no amount of $x_2$ should be produced.

- The maximum profit obtainable is $1033.33 under this arrangement.

A noteworthy point is that the solution indicates a fractional choice, which may not be feasible in a practical situation. This is the limitation of Scipy solver that it cannot solve the so-called integer programming problems. **Other Python packages like `PuLP`** could be an option for such problems. **[See my article here](https://towardsdatascience.com/linear-programming-and-discrete-optimization-with-python-using-pulp-449f3c5f6e99)**.